Firmware/fibre/python/fibre/discovery.py

```diff
@@ -64,25 +67,71 @@ def did_discover_channel(channel):
         """
         try:
             logger.debug("Connecting to device on " + channel._name)
+
+            temp_dir = tempfile.gettempdir()
+
+            cache_miss = True
+            json_crc16 = 0
+
+            # Fetching the json crc to check cache
             try:
-                json_bytes = channel.remote_endpoint_read_buffer(0)
-            except (TimeoutError, ChannelBrokenException):
-                logger.debug("no response - probably incompatible")
-                return
+                json_crc16 = channel.remote_endpoint_operation(0, struct.pack("<I", 0xffffffff), True, 2)
+                json_crc16 = struct.unpack("<H", json_crc16)[0]
+                logger.debug("Device JSON checksum: 0x{:02X} 0x{:02X}".format(json_crc16 & 0xff, (json_crc16 >> 8) & 0xff))
+            except Exception as error:
+                logger.debug("Error fetching JSON CRC, falling back to downloading full JSON")
+                #logger.debug(traceback.format_exc())
+
+            cache_path =  temp_dir + '/fibre_schema_cache_' + str(json_crc16)
+
+            try:
+                json_cache = open(cache_path, 'r+')
+                logger.debug("Found cache file with crc")
+                cache_miss = False
+            except:
+                logger.debug("Cache miss, there is no cache file")
+                cache_miss = True
+         
+            if (cache_miss):
+                # Download the JSON data
+                try:
+                    logger.debug("Getting json schema from USB... this is slow...")
+                    json_bytes = channel.remote_endpoint_read_buffer(0)
+                    json_cache = open(cache_path, 'w+')
+                    json_cache.write(json_bytes.decode("ascii"))
+                    logger.debug("saved json_bytes to file")
+
+                except (TimeoutError, ChannelBrokenException):
+                    logger.debug("no response - probably incompatible")
+                    return
+            else:
+                try:
+                    json_bytes = json_cache.read().encode("ascii")
```

Comment:
This seems redundant? Either store the ascii bytes, or store the decoded string. No point in ever going string -> bytes.

Reply:
I got an error when trying to use the file functions with bytes, it expects a string with encoding

Reply:
okay, just store the string, and arrive at json_string directly (no need to reconstruct some json_bytes)